File: 2-introduction-to-deep-agents/typescript/m1/m1.7_homework_filled.ts
// typescript/m1/m1.7_homework_filled.ts
/**
 * Reference copy of m1.7_homework.ts with TODOs 1 and 2 filled in so you
 * can run it end to end and see what "done" looks like. This is just one
 * possible answer, so yours might be different. Explore!
 */

import { MemorySaver } from "@langchain/langgraph";
import { createDeepAgent } from "deepagents";

import { model } from "../models.js";

const agent = createDeepAgent({
  model,
  checkpointer: new MemorySaver(),
});

// TODO 1 filled in
const threadA = { configurable: { thread_id: "m1-7-homework-thread-a" } };
const threadB = { configurable: { thread_id: "m1-7-homework-thread-b" } };

const FACT_QUESTION = "What's my iguana's name and what does he eat?";

// TODO 2 filled in
async function runScenario() {
  let result = await agent.invoke(
    {
      messages: [
        {
          role: "user",
          content:
            "Remember that my pet iguana is named Steve and only eats dandelion greens.",
        },
      ],
    },
    threadA
  );
  console.log("Thread A, turn 1:");
  console.log(result.messages.at(-1)?.content);

  result = await agent.invoke(
    { messages: [{ role: "user", content: FACT_QUESTION }] },
    threadA
  );
  console.log("\nThread A, turn 2 (same thread, should remember Steve):");
  console.log(result.messages.at(-1)?.content);

  result = await agent.invoke(
    { messages: [{ role: "user", content: FACT_QUESTION }] },
    threadB
  );
  console.log("\nThread B, turn 1 (different thread, should NOT know):");
  console.log(result.messages.at(-1)?.content);

  // A brand-new agent with its own fresh MemorySaver has no history to
  // load, even reused on threadA's exact thread_id.
  const freshAgent = createDeepAgent({ model, checkpointer: new MemorySaver() });
  result = await freshAgent.invoke(
    { messages: [{ role: "user", content: FACT_QUESTION }] },
    threadA
  );
  console.log("\nFresh agent, threadA's thread_id (should NOT know):");
  console.log(result.messages.at(-1)?.content);
  console.log(
    "Same thread_id, but a different MemorySaver instance has no record " +
      "of it: persistence is scoped to the checkpointer, not the thread_id " +
      "string alone."
  );
}

await runScenario();

File: 2-introduction-to-deep-agents/python/m1/m1.7_homework_filled.py
# python/m1/m1.7_homework_filled.py
"""Reference copy of m1.7_homework.py with TODOs 1 and 2 filled in so you
can run it end to end and see what "done" looks like. This is just one
possible answer, so yours might be different. Explore!"""

import warnings

warnings.filterwarnings("ignore", category=DeprecationWarning)

from deepagents import create_deep_agent
from langgraph.checkpoint.memory import MemorySaver

from models import model

agent = create_deep_agent(
    model=model,
    checkpointer=MemorySaver(),
)

# TODO 1 filled in
thread_a = {"configurable": {"thread_id": "m1-7-homework-thread-a"}}
thread_b = {"configurable": {"thread_id": "m1-7-homework-thread-b"}}

FACT_QUESTION = "What's my iguana's name and what does he eat?"


# TODO 2 filled in
def run_scenario():
    result = agent.invoke(
        {"messages": [{"role": "user", "content": "Remember that my pet iguana is named Steve and only eats dandelion greens."}]},
        config=thread_a,
    )
    print("Thread A, turn 1:")
    print(result["messages"][-1].content)

    result = agent.invoke(
        {"messages": [{"role": "user", "content": FACT_QUESTION}]},
        config=thread_a,
    )
    print("\nThread A, turn 2 (same thread, should remember Steve):")
    print(result["messages"][-1].content)

    result = agent.invoke(
        {"messages": [{"role": "user", "content": FACT_QUESTION}]},
        config=thread_b,
    )
    print("\nThread B, turn 1 (different thread, should NOT know):")
    print(result["messages"][-1].content)

    # A brand-new agent with its own fresh MemorySaver has no history to
    # load, even reused on thread_a's exact thread_id.
    fresh_agent = create_deep_agent(model=model, checkpointer=MemorySaver())
    result = fresh_agent.invoke(
        {"messages": [{"role": "user", "content": FACT_QUESTION}]},
        config=thread_a,
    )
    print("\nFresh agent, thread_a's thread_id (should NOT know):")
    print(result["messages"][-1].content)
    print(
        "Same thread_id, but a different MemorySaver instance has no record "
        "of it: persistence is scoped to the checkpointer, not the thread_id "
        "string alone."
    )


run_scenario()
